trino-aws-proxy/src/main/java/io/trino/aws/proxy/server/rest/RequestBuilder.java

```diff
@@ -110,37 +110,54 @@ record BucketAndKey(String bucket, String rawKey) {}
                 requestContent);
     }
 
-    private static RequestContent buildRequestContent(InputStream requestEntityStream, String requestContentType)
+    @SuppressWarnings("SwitchStatementWithTooFewBranches")
+    private static RequestContent buildRequestContent(InputStream requestEntityStream, MultiMap requestHeaders, String requestContentType)
     {
         ContentType contentType = switch (requestContentType) {
             case "aws-chunked" -> ContentType.AWS_CHUNKED;
             case "chunked" -> ContentType.W3C_CHUNKED;
             default -> ContentType.STANDARD;
         };
 
-        Supplier<Optional<byte[]>> bytesSupplier;
-        if (contentType == ContentType.STANDARD) {
-            // memoize the entity bytes so it can be called multiple times
-            bytesSupplier = Suppliers.memoize(() -> {
+        Supplier<Optional<byte[]>> bytesSupplier = switch (contentType) {
+            case STANDARD -> Suppliers.memoize(() -> {
                 try {
                     return Optional.of(toByteArray(requestEntityStream));
                 }
                 catch (IOException e) {
-                    throw new WebApplicationException(Response.Status.BAD_REQUEST);
+                    throw new WebApplicationException(BAD_REQUEST);
                 }
             });
-        }
-        else {
-            // we always stream chunked content. Never load it into memory.
-            bytesSupplier = Optional::empty;
-        }
+
+            default -> Optional::empty;
+        };
+
+        Supplier<Optional<Integer>> contentLengthSupplier = switch (contentType) {
+            case STANDARD -> () -> bytesSupplier.get().map(bytes -> bytes.length);
+
+            case AWS_CHUNKED -> () -> {
+                int contentLength = requestHeaders.getFirst("x-amz-decoded-content-length")
```

Comment:
In addition to that, it appears that you can transfer `aws-chunked` data with a `Transfer-Encoding` of `chunked`

Meaning we'd have double chunking - unclear whether that is even supported, though the docs don't explicitly disallow it

Reply:
> it appears that you can transfer aws-chunked data with a Transfer-Encoding of chunked

I actually tested that at some point. `aws-chunked` is technically normal/spec chunk but with the addition of the signed metadata.

Reply:
> I'm not sure whether this is technically correct. x-amz-decoded-content-length is the size of the object being uploaded, which does not include the extra metadata used for the aws-chunked encoding.

We decode aws-chunked as we read the request (via `AwsChunkedInputStream`). We send a single un-chunked stream to real/remote S3.

Reply:
> > it appears that you can transfer aws-chunked data with a Transfer-Encoding of chunked
> 
> I actually tested that at some point. `aws-chunked` is technically normal/spec chunk but with the addition of the signed metadata.

My question was more on whether (as weird as it would be) it'd be technically feasible to send a request with `Content-Type=aws-chunked` _and_ `Transfer-Encoding=chunked`.

As you say, `aws-chunked` is a superset of the standard `chunked` transfer encoding so I believe this would work correctly.

[The docs](https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-streaming.html) state that:
- You _must_ set `content-type: aws-chunked`, and
- Optionally _may_ set a transfer encoding, which itself may be `chunked` ([docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding))

That would be a weird thing to do but I think the proxy would currently treat it correctly, as it prefers the `content-type` over the `transfer-encoding` value.

I just wonder whether it would make sense to treat content-type and transfer-encoding as separate values - we can do that separately

Reply:
> My question was more on whether (as weird as it would be) it'd be technically feasible to send a request with

I'll add a test for this. There may be an edge case. Jetty/Jersey might possibly try to process the chunked request before we have a chance to work with it.